File: src/utils/text_util.rs
/// Remove ANSI escape sequences from a string.
/// Used when piped output contains cursor/erase sequences.
pub fn strip_ansi(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut chars = s.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\x1b' {
            if chars.peek() == Some(&'[') {
                chars.next();
                for sc in chars.by_ref() {
                    if sc.is_ascii_alphabetic() { break; }
                }
            } else {
                chars.next();
            }
        } else {
            out.push(c);
        }
    }
    out
}

/// Strip trailing `\r` from a line — needed on Windows piped output.
pub fn sanitize_line(line: String) -> String {
    line.trim_end_matches('\r').to_string()
}

/// Find the char-index of the first occurrence of `needle` in `haystack`.
/// Used for column offset detection in tabular CLI output (winget, scoop).
pub fn find_col(haystack: &str, needle: &str) -> Option<usize> {
    let h: Vec<char> = haystack.chars().collect();
    let n: Vec<char> = needle.chars().collect();
    let n_len = n.len();
    if n_len == 0 || h.len() < n_len { return None; }
    for i in 0..=(h.len() - n_len) {
        if h[i..i + n_len] == n[..] { return Some(i); }
    }
    None
}

/// Split "neovim-0.9.5_1" → ("neovim", "0.9.5_1").
/// The version starts at the last hyphen followed by a digit.
/// Used by xbps and apk parsers.
pub fn split_pkg_name_version(pkg: &str) -> (String, String) {
    let chars: Vec<char> = pkg.chars().collect();
    for i in (1..chars.len()).rev() {
        if chars[i - 1] == '-' && chars[i].is_ascii_digit() {
            return (
                chars[..i - 1].iter().collect(),
                chars[i..].iter().collect(),
            );
        }
    }
    (pkg.to_string(), String::new())
}

File: Dotfile-java-TUI/src/main/java/io/github/sampongstarluck/dotfile/service/TextUtil.java
package io.github.sampongstarluck.dotfile.service;

/** Port of {@code ../src/utils/text_util.rs}. */
public final class TextUtil {

    private TextUtil() {}

    /** Remove ANSI escape sequences (ESC[ ... letter, or ESC+1 char). */
    public static String stripAnsi(String s) {
        StringBuilder out = new StringBuilder(s.length());
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == 0x1b) {
                if (i + 1 < s.length() && s.charAt(i + 1) == '[') {
                    i++; // consume '['
                    while (i + 1 < s.length()) {
                        i++;
                        char sc = s.charAt(i);
                        if (Character.isLetter(sc) && sc < 128) break;
                    }
                } else {
                    i++; // consume one char
                }
            } else {
                out.append(c);
            }
        }
        return out.toString();
    }

    /** Strip a trailing {@code \r} — needed on Windows piped output. */
    public static String sanitizeLine(String line) {
        return line.endsWith("\r") ? line.substring(0, line.length() - 1) : line;
    }

    /** char-index (NOT byte) of needle in haystack — for column offset detection. Returns -1 when absent. */
    public static int findCol(String haystack, String needle) {
        return haystack.indexOf(needle);
    }

    /** "neovim-0.9.5_1" -&gt; ["neovim", "0.9.5_1"]; version starts at the last '-' followed by a digit. */
    public static String[] splitPkgNameVersion(String pkg) {
        for (int i = pkg.length() - 1; i >= 1; i--) {
            if (pkg.charAt(i - 1) == '-' && Character.isDigit(pkg.charAt(i))) {
                return new String[] { pkg.substring(0, i - 1), pkg.substring(i) };
            }
        }
        return new String[] { pkg, "" };
    }
}
